### PR TITLE
Enable FileStream length caching with FileShare.Delete

### PR DESCRIPTION
**Pull request by pentp**
`FileShare.Delete` doesn't allow file length to change.
Also cleaned up the `share` parameter from handle based strategy constructors.

### CODE REVIEW

**Comment by stephentoub** [comment]
> Enable FileStream length caching with FileShare.Delete

I must be missing it, but I don't see any changes related to FileShare.Delete...?

**Comment by pentp** [comment]
Previously the check was `_share <= FileShare.Read`, now it's `(share & FileShare.Write) == 0` so it allows length caching if `FileShare.Read | FileShare.Delete` is used to open a file.

File: src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs [hunk 1]
```diff
@@ -13,17 +13,15 @@ internal abstract class OSFileStreamStrategy : FileStreamStrategy
...
         protected readonly SafeFileHandle _fileHandle; // only ever null if ctor throws
         private readonly FileAccess _access; // What file was opened for.
-        private readonly FileShare _share;
 
         protected long _filePosition;
         private long _appendStart; // When appending, prevent overwriting file.
-        private long _length = -1; // When the file is locked for writes on Windows (_share <= FileShare.Read) cache file length in-memory, negative means that hasn't been fetched.
-        private bool _exposedHandle; // created from handle, or SafeFileHandle was used and the handle got exposed
+        private long _length = -1; // When the file is locked for writes on Windows ((share & FileShare.Write) == 0) cache file length in-memory, negative means that hasn't been fetched.
+        private bool _exposedHandle; // created from handle, or SafeFileHandle was used and the handle got exposed, or FileShare.Write was specified when the handle was opened.
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs:20]: It seems that we should change the name of this field. Historically it meant "the handle has been exposed, validate file offset on every async operaiton":

https://github.com/dotnet/runtime/blob/50413eafd5a930dc5f3cce1a7fcba743cbb36c80/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/Net5CompatFileStreamStrategy.cs#L301-L303

Now it means "don't cache file length as it's not safe".

```suggestion
        private bool _lengthCanBeCached;
```

**Review by adamsitnik — COMMENTED** [review]
Overall LGTM, but it would be great to rename the field before we hit the merge button.

@pentp thank you!

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @pentp !

**Pull request merged**